### PR TITLE
Release v0.9.0

### PR DESCRIPTION
**Pull request by angrybayblade**
## Release summary

Version number: 0.9.0

## Release details

Autonomy:
- Updates the on-chain interaction functionalities to wait for the relevant event to make sure the interaction was successful
- Makes the usage of local tendermint chain optional in the deployment setup
- Introduces `autonomy analyse dialogues` command for analysing the dialogue definitions in a skill package
- Introduces `autonomy analyse service` command for checking the deployment readiness of a service
- Introduces `autonomy analyse logs` command for analysing the agent runtime logs
- Adds support for defining custom author name for docker images on the `autonomy build-image` command
- Removes the usage of `MAX_PARTICIPANTS` environment variable from deployment setup
- Adds support for updating the consensus threshold at the runtime using the on-chain metadata

Packages:
- Makes the synchronized database `serializable` and `hashable`
- Backports the `makerDAO multicall2` contract from [`agent-academy-1`](https://github.com/valory-xyz/agent-academy-1) repository
- Simplifies the way the initial height is set on resets. Now, the initial height will always be `0`
- Removes the need for declaring the `payload_attribute` on the round class implementation
- Updates the base rounds to simplify the usage when payloads have multiple attributes
- Adds support for synchronizing the database on registration so that agents can rejoin at any point
- Removes `reset_index` from `AbciApp` implementation
- Adds restrictions on which agents can submit in late message round
- Fixes overlapping blocks by setting unique chain Ids among the tendermint resets
- Updates the rejoin mechanism to
  - Restart tendermint whenever there's a connection drop, from the abci or otherwise
  - Makes sure monitoring is performed even when tendermint is reset by the tendermint server
  - Clean up `timeouts` and `last_timestamp` before trying to restore state received from peers
  - Don't wait for the observation interval in cases when reset is performed for recovery
  - Enables `ACN` communication for rejoining agents, by accepting `requests/responses` from `all_participants`
- Removes consensus from the skills' configurations and replaces the `max_participants` with the length of the `all_participants` list
- Updates the `BaseParams` implementation to enforce minimum values for `observation_interval`
- Updates the `AbciAppDB` implementation to make sure `cross_period_keys` and the database conditions are defined as sets

Tests:
- Adds test coverage for registry contracts
- Adds test for building and running the base autonomy image, the agent runtime image and the tendermint server image
- Re-enables the `fuzzer` tests for the `valory/abci` connection on windows platform

Docs:
- Removes the usage of `max_participants` parameter
- Rearranges the on-chain registration section to match with the autonolas protocol documentation
- Adds a link to the `whitepaper`
- Adds documentation on the deployment readiness checks for a service
- Adds a guide on initializing an empty local packages repository 

Chores:
- Increases `CI` timeout for the tests from `70` minutes to `90` minutes
- Bumps `tomte` to `v0.2.2`

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../main/CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `main` branch (left side), from `develop`
- [ ] I've updated the dependencies versions to the latest, wherever is possible.
- [ ] Lint and unit tests pass locally (please run tests also manually, not only with `tox`)
- [ ] I built the documentation and updated it with the latest changes
- [ ] I've added an item in `HISTORY.md` for this release
- [ ] I bumped the version number in the `__init__.py` file.
- [ ] I published the latest version on TestPyPI and checked that the following command work:
       ```pip install project-name==<version-number> --index-url https://test.pypi.org/simple --force --no-cache-dir --no-deps```
- [ ] After merging the PR, I'll publish the build also on PyPI. Then, I'll make sure the following
      command will work:
      ```pip install project-name-template==<version_number> --force --no-cache-dir --no-deps```  
- [ ] After merging the PR, I'll tag the repo with `v${VERSION_NUMVER}` (e.g. `v0.1.2`)

### CODE REVIEW

File: docs/upgrading.md [hunk 1]
```diff
@@ -7,6 +7,14 @@ Below we describe the additional manual steps required to upgrade between differ
 
 ## `v0.7.0` to `v0.8.0`
 
+Breaking changes
```
Comment (Adamantios) [inline docs/upgrading.md:10]:
```suggestion
## `v0.8.0` to `v0.9.0`

Breaking changes

- the `max_participants` have been removed. 
- the `consensus_threshold` should now be specified in the `setup` params. If `null`, then it is calculated automatically.
```

**Review by Adamantios — APPROVED** [review]

File: autonomy/analyse/service.py [hunk 2]
```diff
@@ -106,6 +106,7 @@
                                     "required": [
                                         "safe_contract_address",
                                         "all_participants",
+                                        "consensus_threshold",
```
Comment (angrybayblade) [inline autonomy/analyse/service.py:109]: Also added this since this is now required

**Review by Adamantios — APPROVED** [review]

File: docs/upgrading.md [hunk 3]
```diff
@@ -5,6 +5,16 @@ Below we describe the additional manual steps required to upgrade between differ
 
 # Open Autonomy
 
+## `v0.8.0` to `v0.9.0`
+
+Breaking changes
+
+- the `max_participants` have been removed.
```
Comment (DavidMinarsch) [inline docs/upgrading.md:12]: removed where? This needs to be a full sentence that's clearly stating where the change occured

File: docs/upgrading.md [hunk 4]
```diff
@@ -5,6 +5,16 @@ Below we describe the additional manual steps required to upgrade between differ
 
 # Open Autonomy
 
+## `v0.8.0` to `v0.9.0`
+
+Breaking changes
+
+- the `max_participants` have been removed. 
+- the `consensus_threshold` should now be specified in the `setup` parameters. If `null`, then it is calculated automatically.
```
Comment (DavidMinarsch) [inline docs/upgrading.md:13]: I would cluster this and the above as subpoints under changes to the configuration

Comment (DavidMinarsch) [inline docs/upgrading.md:13]:
```suggestion
- the `consensus_threshold` should now be specified in the `setup` parameters. If `null`, the recommended setting, then it is calculated automatically from the participants provided.
```

File: docs/upgrading.md [hunk 5]
```diff
@@ -5,6 +5,16 @@ Below we describe the additional manual steps required to upgrade between differ
...
+- The synchronized database is now `serializable` and `hashable`. This means that the data inserted into the database are now enforced to be primitive or non-primitive built-in types only, except for sets and `frozensets`. In essence, the data should be `json serializable`.
```
Comment (DavidMinarsch) [inline docs/upgrading.md:14]: Why except sets and frozensets. What happens with them?

File: docs/upgrading.md [hunk 6]
```diff
@@ -5,6 +5,16 @@ Below we describe the additional manual steps required to upgrade between differ
...
+- The usage of local tendermint chain is optional in the deployment setup, use `-ltm, --local-tm-setup` when building a deployment using `autonomy deploy build` command to include a local tendermint chain setup.
```
Comment (DavidMinarsch) [inline docs/upgrading.md:15]: I mentioned before. `local tendermint chain` is not great terminology. We call it Tendermint Consensus Gadget elsewhere cc @jmoreira-valory

File: docs/upgrading.md [hunk 7]
```diff
@@ -5,6 +5,16 @@ Below we describe the additional manual steps required to upgrade between differ
...
+- The `cross_period_keys` and database conditions needs to be defined as sets
```
Comment (DavidMinarsch) [inline docs/upgrading.md:16]: where and how was it defined before? These upgrading sentences need to be more helpful

File: HISTORY.md [hunk 8]
```diff
@@ -1,5 +1,54 @@
...
+- Introduces `autonomy analyse logs` command for analysing the agent runtime logs
+- Adds support for defining custom author name for docker images on the `autonomy build-image` command
+- Removes the usage of `MAX_PARTICIPANTS` environment variable from deployment setup
+- Adds support for updating the consensus threshold at the runtime using the on-chain metadata
+
+Packages:
+- Makes the synchronized database `serializable` and `hashable`
+- Backports the `makerDAO multicall2` contract from [`agent-academy-1`](https://github.com/valory-xyz/agent-academy-1) repository
+- Simplifies the way the initial height is set on resets. Now, the initial height will always be `0`
+- Removes the need for declaring the `payload_attribute` on the round class implementation
```
Comment (DavidMinarsch) [inline HISTORY.md:19]: This is missing from update guide

File: HISTORY.md [hunk 9]
```diff
@@ -1,5 +1,54 @@
...
+- Updates the base rounds to simplify the usage when payloads have multiple attributes
```
Comment (DavidMinarsch) [inline HISTORY.md:20]: Missing explainer in upgrade guide. Specifically regarding changes to rounds

File: HISTORY.md [hunk 10]
```diff
@@ -1,5 +1,54 @@
...
+- Removes `reset_index` from `AbciApp` implementation
+- Adds restrictions on which agents can submit in late message round
+- Fixes overlapping blocks by setting unique chain Ids among the tendermint resets
+- Updates the rejoin mechanism to
+  - Restart tendermint whenever there's a connection drop, from the abci or otherwise
+  - Makes sure monitoring is performed even when tendermint is reset by the tendermint server
+  - Clean up `timeouts` and `last_timestamp` before trying to restore state received from peers
+  - Don't wait for the observation interval in cases when reset is performed for recovery
+  - Enables `ACN` communication for rejoining agents, by accepting `requests/responses` from `all_participants`
+- Removes consensus from the skills' configurations and replaces the `max_participants` with the length of the `all_participants` list
```
Comment (DavidMinarsch) [inline HISTORY.md:31]: "consensus" -> replace with param in quotes

File: HISTORY.md [hunk 11]
```diff
@@ -1,5 +1,54 @@
...
+- Updates the `BaseParams` implementation to enforce minimum values for `observation_interval`
+- Updates the `AbciAppDB` implementation to make sure `cross_period_keys` and the database conditions are defined as sets
+
+
+Tests:
+- Adds test coverage for registry contracts
+- Adds test for building and running the base autonomy image, the agent runtime image and the tendermint server image
+- Re-enables the `fuzzer` tests for the `valory/abci` connection on windows platform
```
Comment (DavidMinarsch) [inline HISTORY.md:39]: remove quotes from fuzzer

File: HISTORY.md [hunk 12]
```diff
@@ -1,5 +1,54 @@
...
+
+Docs:
+- Removes the usage of `max_participants` parameter
+- Rearranges the on-chain registration section to match with the autonolas protocol documentation
```
Comment (DavidMinarsch) [inline HISTORY.md:43]: Autonolas not autonolas

**Review by DavidMinarsch — CHANGES_REQUESTED** [review]

File: autonomy/fsm/scaffold/templates/components.py [hunk 13]
```diff
@@ -103,11 +103,11 @@ class {AbciApp}(AbciApp[Event]):
...
         final_states: Set[AppState] = {final_states}
         event_to_timeout: EventToTimeout = {{}}
-        cross_period_persisted_keys: List[str] = []
-        db_pre_conditions: Dict[AppState, List[str]] = {{
+        cross_period_persisted_keys: Set[str] = []
+        db_pre_conditions: Dict[AppState, Set[str]] = {{
             {db_pre_conditions}
         }}
-        db_post_conditions: Dict[AppState, List[str]] = {{
+        db_post_conditions: Dict[AppState, Set[str]] = {{
```
Comment (angrybayblade) [inline autonomy/fsm/scaffold/templates/components.py:110]: Just found out the scaffold tool was not up to date with the latest breaking change, is there any other breaking change that should be updated on the scaffold tool? @DavidMinarsch @0xArdi @Adamantios

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**